### DISCUSSION TITLE
[Chapter 05] p82. 5.6.1 묻지 말고 명령하기

### DISCUSSION THREAD

**Discussion by Irisation23** [cowaine/is-my-code-that-strange#30] [Chapter 04-05 (불변, 응집도)] [answered]
> 소프트웨어 설계에는 '묻지 말고, 명령하기(Tell, Don't Ask)' 라는 격언이 있습니다. 이는 다른 객체의 내부 상태(변수)를 기반으로 판단하거나 제어하려고 하지말고, 메서드로 명령해서 객체가 알아서 판단하고 제어하도록 설계하라는 의미입니다.

위의 말에 의거하면 소프트웨어 설계에서 말하는 `Tell, Don't Ask`는 A클래스가 B클래스의 상태를 변경하고 싶다면, A에서 B상태 변경을 호출하지말고 B본인 클래스에서 처리하는 말처럼 보입니다.

- 그렇다면, Setter는 `Tell, Don't Ask` 어긋나지 않는 적절한 방법일까요?

**Comment by snoopingh** [comment]
해석하기 나름일 것 같습니다. Lombok 으로 생성한 단순한 setter 메소드의 형식일 경우를 지양하라는 말 같습니다.
메소드 이름만 setField 이고 메소드 내부에 상세한 판단과 제어에 관련된 로직이 포함되어 있으면 상관없지 않을까요? (ex. 검증 로직)

물론 앞 장에서 의미 있는 이름을 사용하라고 했으니 결국 메소드 명을 바꿔야 하긴 하겠네요.

본문 예제로 생각해보면... (적절한 표현일지는 잘 모르겠네요ㅠ.ㅠ)
setter가 갑옷을 입혀주는 거라면, Tell, Don't Ask에 해당하는 equipArmor 는 갑옷만 건네받고
스스로 판단해서 갑옷을 직접 입는 느낌?? 이랄까요

**Reply by snoopingh** [reply]
답변이 너무 모호해서 추가.

코드 5.31 기준)
A클래스가 B클래스(Equipments) 의 상태를 변경하고 싶을 때, setter 사용 시
equipArmor 메소드의 canCahnge (장비를 착용할 수 있는 상태인지 확인하는 상태) 를 A가 알아야 하겠네요.

이런 경우에는
1. B클래스의 해당 부분에 코드 수정이 있을 때, A 클래스 또한 같이 수정되어야 하는 상황이 생길 수 있어서 응집도가 낮아진다
2. B클래스의 상태값이 외부에 노출되어 있는 것이니 캡슐화가 제대로 되지 않았다.

라고 생각됩니다.

**Comment by firewoody237** [comment, marked as answer]
`Tell Don't Ask`와 `Setter의 미사용`은 다른 맥락인 것 같기도 합니다.
Setter -> 객체의 안정성을 위해 사용하지 않는 것이고
Tell Don't Ask -> 객체를 외부에서 판단하게 하지말고, 내부에서 메시지만 제공한다.

---

`Setter`에 대한 의견을 말씀드리자면..
좋은 코드 관련해서 몇 가지 강의와 책을 읽어봤는데, 대부분 비슷한 내용 같습니다.
Setter를 어떠한 경우에도 쓰지말라는 얘기가 많았습니다.

외부에서 값 변경은 책에서도 안좋게 나와있기 때문에, 생성자를 활용하거나 Setter를 사용하는 대신 신규 객체를 반환해주면 되지 않을까요?

**Comment by Irisation23** [comment]
### 체크 포인트

- 스프링 컨트롤러의 request 메서드 중 암호화 된 내용이 넘어오는 경우(DTO)에는 setter() 메서드의 사용이 불가피 하지 않을까?
  -  `ArgumentResolver` 사용을 권유
